### PR TITLE
Add Viper Mini Support

### LINKED ISSUE #1062: Support for Razer Viper Mini
### Device Info

* Name: Viper Mini
* Product Number: <See under device>
* Model Number: <See under device>

### Device Information
```
Please post the output of
Bus 001 Device 002: ID 1532:008a Razer USA, Ltd Razer Viper Mini
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x008a 
  bcdDevice            2.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1

```

```
 drwxr-xr-x 2 root root 260 Mar 26 10:50 .
drwxr-xr-x 4 root root 680 Mar 26 10:50 ..
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-COMPANY_USB_Device-event-if02 -> ../event6
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-COMPANY_USB_Device-event-kbd -> ../event2
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-COMPANY_USB_Device-if01-event-mouse -> ../event3
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-COMPANY_USB_Device-if01-mouse -> ../mouse0
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-COMPANY_USB_Device-if02-event-kbd -> ../event4
lrwxrwxrwx 1 root root  10 Mar 26 10:50 usb-Corsair_Corsair_VOID_PRO_Wireless_Gaming_Headset-event-if03 -> ../event16
lrwxrwxrwx 1 root root  10 Mar 26 10:50 usb-Razer_Razer_Viper_Mini-event-if01 -> ../event12
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-Razer_Razer_Viper_Mini-event-mouse -> ../event5
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-Razer_Razer_Viper_Mini-if01-event-kbd -> ../event9
lrwxrwxrwx 1 root root  10 Mar 26 10:50 usb-Razer_Razer_Viper_Mini-if02-event-kbd -> ../event13
lrwxrwxrwx 1 root root   9 Mar 26 10:50 usb-Razer_Razer_Viper_Mini-mouse -> ../mouse1
```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>

### PR DESCRIPTION
**Pull request by nightsky30**
Adds Viper Mini Support.  Fixes #1062.

~~**Note:** The Viper Mini supports spectrum and reactive effects.  I added them above the device file sections for Viper and Viper Ultimate mice. The other mice might support these effects as well. But I don't have those mice, so I cannot verify.~~

Tested and Works:
 - [x] Breath Single
 - [x] Breath Dual
 - [x] Breath Random
 - [x] Reactive
 - [x] None
 - [x] Spectrum
 - [x] Static
 - [x] Brightness
 - [x] DPI
 - [x] Polling Rate
 - [x] Serial
 - [x] Firmware

### CODE REVIEW

**Comment by AlexP11223** [comment]
When it will be merged?
I was planning to buy this mouse soon :)

File: driver/razermouse_driver.c [hunk 1]
```diff
@@ -3607,9 +3619,11 @@ static int razer_mouse_probe(struct hid_device *hdev, const struct hid_device_id
...
             CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_led_brightness);
-            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_matrix_effect_breath);
-            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_matrix_effect_static);
-            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_matrix_effect_none);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_spectrum);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_reactive);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_breath);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_static);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_none);
             break;
```
Comment (z3ntu) [inline driver/razermouse_driver.c]: Are you modifying existing devices on purpose? I've seen the crossed out note in the top-most message, please clarify.

Comment (nightsky30) [reply driver/razermouse_driver.c]: According to Razer's documentation all three versions of the Viper should support the same effects. I initially separated the Viper Mini from the other two models as that was the only device I have to test with. It supports spectrum and reactive through my testing. Then in #1164 @troudbal confirmed the Viper also supports the same effects, so I went ahead and included them together.

The Ultimate has admittedly not been tested. I'm fairly certain the Ultimate supports spectrum and reactive just the same as the other two. I could separate the Ultimate out from the Viper and Viper Mini with a switch case and fall-through for the latter two that includes spectrum/reactive if you want.

Comment (z3ntu) [reply driver/razermouse_driver.c]: fine then

File: driver/razermouse_driver.c [hunk 2]
```diff
@@ -3607,9 +3619,11 @@ static int razer_mouse_probe(struct hid_device *hdev, const struct hid_device_id
...
             CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_dpi);
             CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_led_brightness);
-            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_matrix_effect_breath);
-            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_matrix_effect_static);
-            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_logo_matrix_effect_none);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_spectrum);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_reactive);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_breath);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_static);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_scroll_matrix_effect_none);
```
Comment (z3ntu) [inline driver/razermouse_driver.c]: I'm a bit confused why it's scroll here but logo everywhere else?

Comment (nightsky30) [reply driver/razermouse_driver.c]: Because I copy/pasted the wrong lines :smile:

**Comment by Iknash** [comment]
Hey guys! Thanks for this project! Are you planning to merge it soon?

**Comment by NickBelhomme** [comment]
Bump. Don't know why it is not being merged? Can we help with something?

**Comment by AlexP11223** [comment]
> Bump. Don't know why it is not being merged? Can we help with something?

Waiting for code review probably.

You can install this branch yourself. https://github.com/openrazer/openrazer/wiki/Building-a-package

**Comment by AlexP11223** [comment]
Also for anyone considering to buy this mouse, note that OpenRazer currently only supports lighting effects and some limited DPI adjustments. No button remapping, profiles, etc. 
And since Viper Mini has onboard memory according to [the official website](https://www.razer.com/gaming-mice/razer-viper-mini/RZ01-03250100-R3U1), you probably can just set the effects and other settings on Windows.

https://www.reddit.com/r/linux_gaming/comments/iiozgi/psa_currently_there_is_no_easy_way_to_remap/

**Comment by z3ntu** [comment]
@nightsky30 I've merged #1164 so there are a couple conflicts now, could you please resolve them?

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 3]
```diff
@@ -15,6 +15,50 @@
...
+    """
+    EVENT_FILE_REGEX = re.compile(r'.*Razer_Razer_Viper_Mini-if0(1|2)-event-kbd')
+
+    USB_VID = 0x1532
+    USB_PID = 0x008A
+    METHODS = ['get_device_type_mouse', 'max_dpi', 'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate', 'get_logo_brightness', 'set_logo_brightness',
+               # Underglow/Logo use LOGO_LED
+               'set_logo_static_naga_hex_v2', 'set_logo_spectrum_naga_hex_v2', 'set_logo_none_naga_hex_v2', 'set_logo_reactive_naga_hex_v2', 'set_logo_breath_random_naga_hex_v2', 'set_logo_breath_single_naga_hex_v2', 'set_logo_breath_dual_naga_hex_v2',
+               # Can set Underglow/LOGO with custom
+               'set_custom_effect', 'set_key_row']
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: If the device has custom effect support, it needs HAS_MATRIX = True and the matrix dimensions set

**Review by z3ntu — APPROVED** [review]
LGTM otherwise

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 3 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/mouse.py]: I've removed the references to custom effect support here as well as in the fake driver

**Comment by adrianro9224** [comment]
Hi guys, i buy this mouse and have Ubuntu as OS, the daemon version is 2.8.0 but the device is not recognized :(.

**Comment by bohrshaw** [comment]
@adrianro9224 This PR is not in a release yet.

**Comment by adrianro9224** [comment]
Hi @bohrshaw, sorry now i understand, thanks for your help!

**Pull request merged**